### PR TITLE
Update Application Signals Processor to use extension if available

### PR DESCRIPTION
**Pull request by musa-asad**
# Description of the issue
A new [Kubernetes Metadata Extension](https://github.com/aws/amazon-cloudwatch-agent/tree/main/extension/k8smetadata) was implemented for the CloudWatch Agent, which contains mappings for Kubernetes metadata containing the pod name, namespace, and node name of a Kubernetes application. This extension was only configured for the [Entity Processor](https://github.com/aws/amazon-cloudwatch-agent/tree/main/plugins/processors/awsentity), but it can be migrated to also be used for the [Application Signals Processor](https://github.com/aws/amazon-cloudwatch-agent/tree/main/plugins/processors/awsapplicationsignals) so that there is one singleton to get shared Kubernetes metadata from, minimizing redundant API calls and caching.

# Description of changes
- Updated `k8smetadata` extension to include `serviceWatcher`.
  - Created `serviceWatcher` to update and cache Service objects.
  - Added `GetPodMetadataFromServiceAndNamespace` and `GetServiceAndNamespaceFromClusterIP` methods.
  - Update `README.md` to mention inclusion of Service objects.
- Completely migrated watchers from `awsapplicationsignals` to `k8sCommon`, while ensuring `workload@namespace` is formatted as `{workload, namespace, node}` instead.
- Updated `kubernetes.go` to utilize the `k8smetadata` extension if available to retrieve necessary mappings for Application Signals `RemoteService` metadata.
- Add `k8smetadata` extension to Application Signals pipeline for Kubernetes contexts.
- Added unit tests for new functionality.

# License
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

# Tests
## Integration Tests
Workflow Status
[View Workflow Run #14234895874](https://github.com/aws/amazon-cloudwatch-agent/actions/runs/14234895874)

## Application Signals E2E Tests
Workflow Status
[View Workflow Run #14234895877](https://github.com/aws/amazon-cloudwatch-agent/actions/runs/14234895877)

## Manual Tests
Deployed a sample application with https://aws-otel.github.io/docs/getting-started/adot-eks-add-on/sample-app.

**Default**
```
2025-04-03T07:09:02Z D! Fallback shared config file(s): [/.aws/credentials]                                                                                       │
│ 2025-04-03T07:09:02Z D! Successfully created credential sessions                                                                                                  │
│ 2025-04-03T07:09:02Z D! Using credential XXXXXXXXXXXXX from EC2RoleProvider                                                                                │
│ W0403 07:09:02.150959       1 client_config.go:659] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.            │
│ I0403 07:09:03.157439       1 shared_informer.go:313] Waiting for caches to sync for endpointSliceWatcher                                                         │
│ I0403 07:09:03.258078       1 shared_informer.go:320] Caches are synced for endpointSliceWatcher                                                                  │
│ 2025-04-03T07:09:03Z I! {"caller":"k8sclient/endpointslicewatcher.go:88","msg":"endpointSliceWatcher: Cache synced","kind":"extension","name":"k8smetadata"}      │
│ I0403 07:09:03.258232       1 shared_informer.go:313] Waiting for caches to sync for serviceWatcher                                                               │
│ I0403 07:09:03.258247       1 shared_informer.go:320] Caches are synced for serviceWatcher                                                                        │
│ 2025-04-03T07:09:03Z I! {"caller":"k8sclient/servicewatcher.go:66","msg":"serviceWatcher: Cache synced","kind":"extension","name":"k8smetadata"}                  │
│ 2025-04-03T07:09:03Z I! {"caller":"awsproxy@v0.115.0/extension.go:39","msg":"X-Ray proxy server started on 0.0.0.0:2000","kind":"extension","name":"awsproxy/appl │
│ ication_signals"}                                                                                                                                                 │
│ 2025-04-03T07:09:03Z I! {"caller":"resolver/kubernetes.go:167","msg":"k8smetadata extension is present; watchers are skipped","kind":"processor","name":"awsappli │
│ cationsignals","pipeline":"traces/application_signals"}


│ 2025-04-03T07:09:39Z D! {"caller":"k8smetadata/extension.go:96","msg":"GetPodMetadataFromPodIP: found metadata","kind":"extension","name":"k8smetadata","ip":"XXX│
│ .XX.XX.XXX","workload":"sample-app","namespace":"default","node":"ip-XXX-XX-XX-XX.us-west-2.compute.internal"}

2025-04-03T07:14:53Z D! {"caller":"k8sclient/endpointslicewatcher.go:218","msg":"Received EndpointSlice Update","kind":"extension","name":"k8smetadata","oldSlice │
│ UID":"d67b19bf-e207-4c4c-b53e-667712095508","newSliceUID":"d67b19bf-e207-4c4c-b53e-667712095508","name":"sample-app-sv52w","namespace":"default"}                 │
│ 2025-04-03T07:14:53Z D! {"caller":"k8sclient/endpointslicewatcher.go:118","msg":"Processing endpoint","kind":"extension","name":"k8smetadata","podName":"sample-a │
│ pp-5c6465d5bb-567ch","namespace":"default","nodeName":"ip-XXX-XX-XX-XX.us-west-2.compute.internal"}                                                               │
│ 2025-04-03T07:14:53Z D! {"caller":"k8sclient/endpointslicewatcher.go:276","msg":"Finished handling EndpointSlice Update","kind":"extension","name":"k8smetadata", │
│ "sliceUID":"d67b19bf-e207-4c4c-b53e-667712095508"}                                                                                                                │
│ 2025-04-03T07:14:54Z D! {"caller":"k8sclient/endpointslicewatcher.go:218","msg":"Received EndpointSlice Update","kind":"extension","name":"k8smetadata","oldSlice │
│ UID":"d67b19bf-e207-4c4c-b53e-667712095508","newSliceUID":"d67b19bf-e207-4c4c-b53e-667712095508","name":"sample-app-sv52w","namespace":"default"}                 │
│ 2025-04-03T07:14:54Z D! {"caller":"k8sclient/endpointslicewatcher.go:276","msg":"Finished handling EndpointSlice Update","kind":"extension","name":"k8smetadata", │
│ "sliceUID":"d67b19bf-e207-4c4c-b53e-667712095508"}                                                                                                                │                                                                                                  │
│ 2025-04-03T07:14:55Z D! {"caller":"k8sclient/endpointslicewatcher.go:118","msg":"Processing endpoint","kind":"extension","name":"k8smetadata","podName":"sample-a │
│ pp-5c6465d5bb-fbvgg","namespace":"default","nodeName":"ip-XXX-XX-XX-XX.us-west-2.compute.internal"}
```

**PodWatcher**
```
2025-04-03T07:26:14Z D! Using credential XXXXXXXXXXXXX from EC2RoleProvider                                                                                │
│ W0403 07:26:14.965178       1 client_config.go:659] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.            │
│ I0403 07:26:21.965848       1 shared_informer.go:313] Waiting for caches to sync for endpointSliceWatcher                                                         │
│ I0403 07:26:22.066965       1 shared_informer.go:320] Caches are synced for endpointSliceWatcher                                                                  │
│ 2025-04-03T07:26:22Z I! {"caller":"k8sclient/endpointslicewatcher.go:88","msg":"endpointSliceWatcher: Cache synced","kind":"extension","name":"k8smetadata"}      │
│ I0403 07:26:22.067190       1 shared_informer.go:313] Waiting for caches to sync for serviceWatcher                                                               │
│ I0403 07:26:22.067199       1 shared_informer.go:320] Caches are synced for serviceWatcher                                                                        │
│ 2025-04-03T07:26:22Z I! {"caller":"k8sclient/servicewatcher.go:66","msg":"serviceWatcher: Cache synced","kind":"extension","name":"k8smetadata"}                  │
│ 2025-04-03T07:26:22Z I! {"caller":"awsproxy@v0.115.0/extension.go:39","msg":"X-Ray proxy server started on 0.0.0.0:2000","kind":"extension","name":"awsproxy/appl │
│ ication_signals"}                                                                                                                                                 │
│ 2025-04-03T07:26:22Z I! {"caller":"resolver/kubernetes.go:105","msg":"APP_SIGNALS_USE_LIST_POD=true; setting up Pod & Service watchers, ignoring extension","kind │
│ ":"processor","name":"awsapplicationsignals","pipeline":"metrics/application_signals"}                                                                            │
│ W0403 07:26:22.081700       1 client_config.go:659] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.            │
│ I0403 07:26:27.083774       1 shared_informer.go:313] Waiting for caches to sync for podWatcher                                                                   │
│ I0403 07:26:27.184963       1 shared_informer.go:320] Caches are synced for podWatcher                                                                            │
│ 2025-04-03T07:26:27Z I! {"caller":"resolver/podwatcher.go:165","msg":"podWatcher: Cache synced","kind":"processor","name":"awsapplicationsignals","pipeline":"met │
│ rics/application_signals"}

2025-04-03T07:28:51Z D! {"caller":"resolver/podwatcher.go:146","msg":"list and watch for pods: UPDATE sample-app-5c6465d5bb-j8rxf","kind":"processor","name":"aws │
│ applicationsignals","pipeline":"traces/application_signals"}                                                                                                      │
│ 2025-04-03T07:28:51Z I! {"caller":"transport/http2_server.go:666","msg":"[transport] [server-transport 0xc001940d00] Closing: EOF","grpc_log":true}               │
│ 2025-04-03T07:28:51Z I! {"caller":"transport/controlbuf.go:577","msg":"[transport] [server-transport 0xc001940d00] loopyWriter exiting with error: transport clos │
│ ed by client","grpc_log":true}                                                                                                                                    │
│ 2025-04-03T07:28:52Z D! {"caller":"resolver/podwatcher.go:146","msg":"list and watch for pods: UPDATE sample-app-5c6465d5bb-fbvgg","kind":"processor","name":"aws │
│ applicationsignals","pipeline":"traces/application_signals"}                                                                                                      │
│ 2025-04-03T07:28:52Z D! {"caller":"k8sclient/endpointslicewatcher.go:218","msg":"Received EndpointSlice Update","kind":"extension","name":"k8smetadata","oldSlice │
│ UID":"d67b19bf-e207-4c4c-b53e-667712095508","newSliceUID":"d67b19bf-e207-4c4c-b53e-667712095508","name":"sample-app-sv52w","namespace":"default"}                 │
│ 2025-04-03T07:28:52Z D! {"caller":"k8sclient/endpointslicewatcher.go:276","msg":"Finished handling EndpointSlice Update","kind":"extension","name":"k8smetadata", │
│ "sliceUID":"d67b19bf-e207-4c4c-b53e-667712095508"}                                                                                                                │
│ 2025-04-03T07:28:52Z D! {"caller":"resolver/podwatcher.go:151","msg":"list and watch for pods: DELETE sample-app-5c6465d5bb-fbvgg","kind":"processor","name":"aws │
│ applicationsignals","pipeline":"traces/application_signals"}                                                                                                      │
│ 2025-04-03T07:28:52Z D! {"caller":"resolver/podwatcher.go:91","msg":"deleting pod ip from cache","kind":"processor","name":"awsapplicationsignals","pipeline":"tr │
│ aces/application_signals","podNetwork":"XXX.XX.XX.XX"}                                                                                                            │
│ 2025-04-03T07:28:52Z D! {"caller":"resolver/podwatcher.go:98","msg":"decrementing pod count","kind":"processor","name":"awsapplicationsignals","pipeline":"traces │
│ /application_signals","workload":"sample-app@default","podCount":1}                                                                                               │
│ 2025-04-03T07:28:53Z D! {"caller":"k8sclient/endpointslicewatcher.go:118","msg":"Processing endpoint","kind":"extension","name":"k8smetadata","podName":"sample-a │
│ pp-5c6465d5bb-j8rxf","namespace":"default","nodeName":"ip-XXX-XX-XX-XX.us-west-2.compute.internal"}
```

**EMF Log (Default)**
```
{
    "EC2.AutoScalingGroup": "eks-nodegroup-bacafd1d-d73d-0336-dc96-6be3ea1d7ac3",
    "EC2.InstanceId": "i-030007521bea0b48e",
    "EKS.Cluster": "app-sig-ext",
    "Environment": "eks:app-sig-ext/default",
    "Host": "ip-XXX-XX-XX-XX.us-west-2.compute.internal",
    "K8s.Namespace": "default",
    "K8s.Node": "i-030007521bea0b48e",
    "K8s.Pod": "sample-app-85644b64f-n78zt",
    "K8s.Workload": "sample-app",
    "Operation": "GET /outgoing-http-call",
    "PlatformType": "AWS::EKS",
    "RemoteOperation": "GET /",
    "RemoteService": "aws.amazon.com:443",
    "Service": "sample-app",
    "Telemetry.Agent": "CWAgent/Unknown",
    "Telemetry.SDK": "opentelemetry-java-instrumentation,2.0.1-aws,java,Auto",
    "Telemetry.Source": "ClientSpan",
    "Version": "1",
    "_aws": {
        "CloudWatchMetrics": [
            {
                "Namespace": "ApplicationSignals",
                "Dimensions": [
                    [
                        "Environment",
                        "Operation",
                        "RemoteOperation",
                        "RemoteService",
                        "Service"
                    ],
                    [
                        "Environment",
                        "RemoteService",
                        "Service"
                    ],
                    [
                        "Environment",
                        "RemoteOperation",
                        "RemoteService",
                        "Service"
                    ],
                    [
                        "RemoteService"
                    ]
                ],
                "Metrics": [
                    {
                        "Name": "Latency",
                        "Unit": "Milliseconds",
                        "StorageResolution": 60
                    },
                    {
                        "Name": "Error",
                        "Unit": "",
                        "StorageResolution": 60
                    },
                    {
                        "Name": "Fault",
                        "Unit": "",
                        "StorageResolution": 60
                    }
                ]
            }
        ],
        "Timestamp": 1743665989169
    },
    "aws.log.group.names": "/aws/containerinsights/app-sig-ext/application",
    "Error": {
        "Values": [
            0
        ],
        "Counts": [
            2
        ],
        "Max": 0,
        "Min": 0,
        "Count": 2,
        "Sum": 0
    },
    "Fault": {
        "Values": [
            0
        ],
        "Counts": [
            2
        ],
        "Max": 0,
        "Min": 0,
        "Count": 2,
        "Sum": 0
    },
    "Latency": {
        "Values": [
            164.65317692705105,
            100.59041552234048
        ],
        "Counts": [
            1,
            1
        ],
        "Max": 165.092684,
        "Min": 100.836979,
        "Count": 2,
        "Sum": 265.929663
    }
}
```

**EKS Cluster app-sig-ext > Namespace default > Workload sample-app**

> Note: I noticed the cache update logs were present in a single agent only. Is this expected?

# Requirements
_Before commit the code, please do the following steps._
1. Run `make fmt` and `make fmt-sh`
3. Run `make lint`

### CODE REVIEW

File: internal/k8sCommon/k8sclient/endpointslicewatcher.go [hunk 1]
```diff
@@ -76,6 +80,7 @@ func (w *EndpointSliceWatcher) Run(stopCh chan struct{}) {
 	go w.informer.Run(stopCh)
 }
 
+// WaitForCacheSync blocks until the cache is synchronized, or the stopCh is closed.
 func (w *EndpointSliceWatcher) WaitForCacheSync(stopCh chan struct{}) {
 	if !cache.WaitForNamedCacheSync("endpointSliceWatcher", stopCh, w.informer.HasSynced) {
 		w.logger.Error("timed out waiting for endpointSliceWatcher cache to sync")
```
Comment (musa-asad) [inline internal/k8sCommon/k8sclient/endpointslicewatcher.go:86]: FYI, we no longer use fatal here, which might be an issue.

File: extension/k8smetadata/extension.go [hunk 2]
```diff
@@ -30,6 +30,7 @@ type KubernetesMetadata struct {
 	ready                atomic.Bool
 	safeStopCh           *k8sclient.SafeChannel
 	endpointSliceWatcher *k8sclient.EndpointSliceWatcher
+	serviceWatcher       *k8sclient.ServiceWatcher
```
Comment (musa-asad) [inline extension/k8smetadata/extension.go:33]: Curious if we should have some flag to enable the `serviceWatcher` since we don't need it for Entity?

**Comment by lisguo** [comment]
>    Note: I noticed the cache update logs were present in a single agent only. Is this expected?

Seems odd. All agents should have their own cache in memory and listening on endpointslice/service changes... I'd expect all of them to modify the cache to stay in sync.


Also for testing this, do you have multiple apps on the cluster talking to each other? If you have the sample pet clinic app you should see different workloads talking to the frontend which will populate the RemoteService to a workload name on the cluster itself. In your description your RemoteService is  `"RemoteService": "aws.amazon.com:443",`. We should probably test with the pet clinic app or something with multiple workloads.

File: translator/tocwconfig/sampleConfig/appsignals_and_eks_config.yaml [hunk 3]
```diff
@@ -306,6 +306,7 @@ extensions:
         kubernetes_mode: EKS
         mode: ec2
         region: us-east-1
+    k8smetadata: {}
```
Comment (lisguo) [inline translator/tocwconfig/sampleConfig/appsignals_and_eks_config.yaml:309]: nit: This is probably fine for now, but it would be cleaner if we had a config in the k8smetadata extension to specify what API objects to watch on so that this extension is more generic and can extend to multiple use cases.

```
extension:
    k8smetadata:
        objects:
        - endpointslices
        - services

```
Similar to the kubernetes objects receiver: https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/k8sobjectsreceiver

But for the context of CWAgent this is fine. It's non blocking

File: extension/k8smetadata/extension.go (hunk 2 above)
Comment (lisguo) [reply extension/k8smetadata/extension.go:33]: I think we would want to be consistent with appsignals. Any entity we emit should match what app signals will produce so that customers will see 1 service in the topology map

File: internal/k8sCommon/k8sclient/endpointslicewatcher.go [hunk 4]
```diff
@@ -15,13 +15,15 @@ import (
...
 //  1. ip/ip:port -> {"workload", "namespace", "node"}
+//  2. service@namespace -> {"workload", "namespace", "node"}
 type EndpointSliceWatcher struct {
-	logger          *zap.Logger
-	informer        cache.SharedIndexInformer
-	IPToPodMetadata *sync.Map // key: "ip" or "ip:port", val: {"workload", "namespace", "node"}
+	logger               *zap.Logger
+	informer             cache.SharedIndexInformer
+	IPToPodMetadata      *sync.Map // key: "ip" or "ip:port", val: PodMetadata
+	ServiceToPodMetadata *sync.Map // key: "serviceName@namespace", val: PodMetadata
```
Comment (lisguo) [inline internal/k8sCommon/k8sclient/endpointslicewatcher.go:23]: Why is IPToPodMetadata and ServiceToPodMetadata exported? We could have getter funcs

File: internal/k8sCommon/k8sclient/endpointslicewatcher.go (hunk 1 above)
Comment (lisguo) [reply internal/k8sCommon/k8sclient/endpointslicewatcher.go:86]: Remind me -- does this function get called again later? So if this fails the first time we try to sync the cache again?

File: internal/k8sCommon/k8sclient/endpointslicewatcher.go [hunk 5]
```diff
@@ -15,13 +15,15 @@ import (
```
Comment (lisguo) [inline internal/k8sCommon/k8sclient/endpointslicewatcher.go:15]: Note to self -- I am cross referencing these changes with what's already implemented in appsignals processor: https://github.com/aws/amazon-cloudwatch-agent/blob/main/plugins/processors/awsapplicationsignals/internal/resolver/endpointslicewatcher.go

File: internal/k8sCommon/k8sclient/servicetoworkload.go [hunk 6]
```diff
@@ -11,16 +11,16 @@ import (
...
 	serviceAndNamespaceToSelectors *sync.Map
 	workloadAndNamespaceToLabels   *sync.Map
 	serviceToWorkload              *sync.Map
 	logger                         *zap.Logger
 	deleter                        Deleter
 }
 
-func newServiceToWorkloadMapper(serviceAndNamespaceToSelectors, workloadAndNamespaceToLabels, serviceToWorkload *sync.Map, logger *zap.Logger, deleter Deleter) *serviceToWorkloadMapper {
-	return &serviceToWorkloadMapper{
+func NewServiceToWorkloadMapper(serviceAndNamespaceToSelectors, workloadAndNamespaceToLabels, serviceToWorkload *sync.Map, logger *zap.Logger, deleter Deleter) *ServiceToWorkloadMapper {
```
Comment (lisguo) [inline internal/k8sCommon/k8sclient/servicetoworkload.go:22]: Are we exporting a lot of these functions and variables just for the appsignals tests? Instead can we provide a mock for appsignals to use? 

Ideally we should be testing the private functions in the k8smetadata extension package directly, and then expose a mock'd struct for other packages to test against

File: internal/k8sCommon/k8sclient/endpointslicewatcher.go [hunk 7]
```diff
@@ -114,39 +121,52 @@ func (w *EndpointSliceWatcher) extractEndpointSliceKeyValuePairs(slice *discv1.E
...
 						})
 					}
 				}
 			}
-		}
 
+			// Build service name -> PodMetadata pair from the first pod
+			if isFirstPod {
+				isFirstPod = false
+				if svcName != "" {
```
Comment (nathalapooja) [inline internal/k8sCommon/k8sclient/endpointslicewatcher.go:161]: Can we also check ns != ""?

Comment (nathalapooja) [reply internal/k8sCommon/k8sclient/endpointslicewatcher.go:161]: Also add a test case if svcName is empty or ns is empty

File: plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go [hunk 8]
```diff
@@ -83,87 +89,120 @@ func jitterSleep(seconds int) {
...
-		}
+		// Check environment for "list pods" approach
+		useListPod := (os.Getenv(appSignalsUseListPod) == "true")
 
-		clientset, err := kubernetes.NewForConfig(config)
-		if err != nil {
-			logger.Fatal("Failed to create kubernetes client", zap.Error(err))
-		}
+		if useListPod {
+			logger.Info("APP_SIGNALS_USE_LIST_POD=true; setting up Pod & Service watchers, ignoring extension")
```
Comment (nathalapooja) [inline plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go:96]: This says ignoring extension, Are we starting the extension and not using in this case?

File: plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go [hunk 9]
```diff
@@ -177,30 +216,61 @@ func (e *kubernetesResolver) Stop(_ context.Context) error {
...
+			return "", "", errors.New("extension not found (unexpected)")
+		}
+		pm := ext.GetPodMetadataFromPodIP(ip)
+		if pm.Workload != "" {
+			return pm.Workload, pm.Namespace, nil
+		}
 
+		if svcKeyVal := ext.GetServiceAndNamespaceFromClusterIP(ip); svcKeyVal != "" {
+			sm := ext.GetPodMetadataFromServiceAndNamespace(svcKeyVal)
+			if sm.Workload != "" {
```
Comment (nathalapooja) [inline plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go:232]: Why we are not validating the namespace being empty?

File: translator/tocwconfig/sampleConfig/appsignals_and_eks_config.yaml (hunk 3 above)
Comment (nathalapooja) [reply translator/tocwconfig/sampleConfig/appsignals_and_eks_config.yaml:309]: This will be interesting addition to reduce resource usage

File: internal/k8sCommon/k8sclient/endpointslicewatcher.go (hunk 1 above)
Comment (lisguo) [reply internal/k8sCommon/k8sclient/endpointslicewatcher.go:86]: why should we crash the agent if it's unable to send entity? I know we crashed for appsignals -- but this is no longer an appsignals specific thing. IMO it's not worth killing the agent just because we aren't able to resolve a RemoteService name or a workload name

File: extension/k8smetadata/extension.go (hunk 2 above)
Comment (lisguo) [reply extension/k8smetadata/extension.go:33]: but what if a user later enables appsignals? Wouldnt we want the same service name resolution?

File: internal/k8sCommon/k8sclient/endpointslicewatcher.go (hunk 1 above)
Comment (musa-asad) [reply internal/k8sCommon/k8sclient/endpointslicewatcher.go:86]: WaitForCacheSync doesn't get called again, rather it runs in the background.

File: extension/k8smetadata/extension.go (hunk 2 above)
Comment (musa-asad) [reply extension/k8smetadata/extension.go:33]: Ok, we will have to update our Entity Processor to use the same logic then.

File: plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go (hunk 9 above)
Comment (musa-asad) [reply plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go:232]: I believe we do with `svcKeyVal != ""`?

File: plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go (hunk 8 above)
Comment (musa-asad) [reply plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go:96]: It isn't being started. I can update the logging.

File: internal/k8sCommon/k8sclient/endpointslicewatcher.go (hunk 7 above)
Comment (musa-asad) [reply internal/k8sCommon/k8sclient/endpointslicewatcher.go:161]: I can add a check for that.

File: internal/k8sCommon/k8sclient/endpointslicewatcher.go [hunk 10]
```diff
@@ -114,39 +121,52 @@ func (w *EndpointSliceWatcher) extractEndpointSliceKeyValuePairs(slice *discv1.E
...
+							value:     fullWl,
+							isService: false,
 						})
 					}
 				}
 			}
-		}
 
+			// Build service name -> PodMetadata pair from the first pod
+			if isFirstPod {
```
[the remark this answers is not in the document]
Comment (musa-asad) [reply internal/k8sCommon/k8sclient/endpointslicewatcher.go:159]: We build the Service -> PodMetadata pair for the first pod it finds in the given endpoints of any EndpointSlice. Not necessarily the first EndpointSlice.

File: extension/k8smetadata/extension.go [hunk 11]
```diff
@@ -78,22 +82,59 @@ func (e *KubernetesMetadata) Shutdown(_ context.Context) error {
...
+		e.logger.Debug("GetPodMetadataFromService: no service@namespace provided")
+		return k8sclient.PodMetadata{}
+	}
+	pm, ok := e.endpointSliceWatcher.ServiceToPodMetadata.Load(svcAndNS)
+	if !ok {
+		e.logger.Debug("GetPodMetadataFromService: no mapping found", zap.String("svcAndNS", svcAndNS))
+		return k8sclient.PodMetadata{}
+	}
+	metadata := pm.(k8sclient.PodMetadata)
+	e.logger.Debug("GetPodMetadataFromService: found metadata",
```
[the remark this answers is not in the document]
Comment (musa-asad) [reply extension/k8smetadata/extension.go:116]: Good catch, forgot to fully update while renaming.

File: extension/k8smetadata/extension.go [hunk 12]
```diff
@@ -78,22 +82,59 @@ func (e *KubernetesMetadata) Shutdown(_ context.Context) error {
...
 	)
 	return metadata
 }
+
+func (e *KubernetesMetadata) GetPodMetadataFromServiceAndNamespace(svcAndNS string) k8sclient.PodMetadata {
+	if svcAndNS == "" {
+		e.logger.Debug("GetPodMetadataFromService: no service@namespace provided")
+		return k8sclient.PodMetadata{}
+	}
+	pm, ok := e.endpointSliceWatcher.ServiceToPodMetadata.Load(svcAndNS)
```
[the remark this answers is not in the document]
Comment (musa-asad) [reply extension/k8smetadata/extension.go:110]: It would be more accurate to do so. I can update.

File: extension/k8smetadata/extension_test.go [hunk 13]
```diff
@@ -59,6 +59,64 @@ func TestKubernetesMetadata_GetPodMetadata_Incomplete(t *testing.T) {
 		endpointSliceWatcher: esw,
 	}
 
-	got := kMeta.GetPodMetadata(testIP)
+	got := kMeta.GetPodMetadataFromPodIP(testIP)
 	assert.Equal(t, expected, got, "GetPodMetadata should return the stored incomplete PodMetadata for IP %s", testIP)
 }
+
+func TestKubernetesMetadata_GetPodMetadataFromService(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (musa-asad) [reply extension/k8smetadata/extension_test.go:66]: Thanks, forgot to update.

File: internal/k8sCommon/k8sclient/servicetoworkload.go (hunk 6 above)
Comment (musa-asad) [reply internal/k8sCommon/k8sclient/servicetoworkload.go:22]: This one in particular is exported for `kubernetes.go`, but it is also being used in the Application Signals tests. It would be exported regardless. Though, I can make a mock regardless of this.

File: internal/k8sCommon/k8sclient/endpointslicewatcher.go (hunk 4 above)
Comment (musa-asad) [reply internal/k8sCommon/k8sclient/endpointslicewatcher.go:23]: True.

File: translator/tocwconfig/sampleConfig/appsignals_and_eks_config.yaml (hunk 3 above)
Comment (musa-asad) [reply translator/tocwconfig/sampleConfig/appsignals_and_eks_config.yaml:309]: I will do this.

File: plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go [hunk 14]
```diff
@@ -83,87 +89,120 @@ func jitterSleep(seconds int) {
...
-			// wait for caches to sync (for once) so that clients knows about the pods and services in the cluster
-			svcWatcher.waitForCacheSync(safeStopCh.ch)
-			endptSliceWatcher.waitForCacheSync(safeStopCh.ch)
+		// 2) If not using listPod, check if extension is present
+		ext := k8smetadata.GetKubernetesMetadata()
+		if ext != nil {
+			// We skip all watchers (the extension has them).
+			logger.Info("k8smetadata extension is present")
 
 			instance = &kubernetesResolver{
```
[the remark this answers is not in the document]
Comment (musa-asad) [reply plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go:154]: The extension is available throughout the entire agent process.

File: extension/k8smetadata/extension.go [hunk 15]
```diff
@@ -59,13 +60,16 @@ func (e *KubernetesMetadata) Start(_ context.Context, _ component.Host) error {
...
+	e.serviceWatcher.Run(e.safeStopCh.Ch)
 
 	e.endpointSliceWatcher.WaitForCacheSync(e.safeStopCh.Ch)
+	e.serviceWatcher.WaitForCacheSync(e.safeStopCh.Ch)
 
-	e.logger.Debug("EndpointSlice cache synced, extension fully started")
+	e.logger.Debug("EndpointSlice and Service cache synced, extension fully started")
 	e.ready.Store(true)
 
 	return nil
```
[the remark this answers is not in the document]
Comment (musa-asad) [reply extension/k8smetadata/extension.go:81]: Throughout the entire agent process.

File: internal/k8sCommon/k8sclient/servicetoworkload.go (hunk 6 above)
Comment (musa-asad) [reply internal/k8sCommon/k8sclient/servicetoworkload.go:22]: Thinking back, why would we mock when we can test the main function itself?

File: extension/k8smetadata/extension.go (hunk 2 above)
Comment (musa-asad) [reply extension/k8smetadata/extension.go:33]: How would we set this up? We currently rely on getting an application pod's IP and then resolving entity from that. Application Signals does the same thing, but it also creates a resolution from the cluster IP -> service.

Comment (lisguo) [reply extension/k8smetadata/extension.go:33]: Okay, I think I see your point -- appsignals only uses the servicewatcher to resolve the clusterIP to the service name. But this shouldn't happen for the pod ip context

**Review by lisguo — APPROVED** [review]
LGTM -- let's just make sure we test appsignals with a petclinic sample app or something similar

**Review by nathalapooja — APPROVED** [review]
